### PR TITLE
Support using /show to show Spotify embeds in chat

### PR DESCRIPTION
**Pull request by sirKiraUzumaki**
works with both spotify link ("spotify:track:SongId") and song link ("https://open.spotify.com/track/SongId")

### CODE REVIEW

File: server/chat-commands/info.ts [hunk 1]
```diff
@@ -2569,10 +2572,16 @@ export const commands: ChatCommands = {
...
+		}
+		const Spotify = new SpotifyInterface();
+		const YouTube = new YoutubeInterface();
+		if (Spotify.linkRegex.test(request.link) || YouTube.linkRegex.test(request.link)) {
+			if (Spotify.linkRegex.test(request.link)) {
+				buf = Spotify.generateSongDisplay(request.link);
+			} else {
+				buf = await YouTube.generateVideoDisplay(request.link);
+				if (!buf) return this.errorReply('Could not get YouTube video');
+			}
```
Comment (AnnikaCodes) [inline server/chat-commands/info.ts:2584]:
```suggestion
		if (Spotify.linkRegex.test(request.link)) {
			buf = Spotify.generateSongDisplay(request.link);
		} else if (YouTube.linkRegex.test(request.link)) {
			buf = await YouTube.generateVideoDisplay(request.link);
			if (!buf) return this.errorReply('Could not get YouTube video');
		}
```

Comment (AnnikaCodes) [reply server/chat-commands/info.ts:2584]: Do we need an error for being unable to fetch a Spotify song?

File: server/chat-commands/info.ts [hunk 2]
```diff
@@ -2622,12 +2631,16 @@ export const commands: ChatCommands = {
...
-			buf = await YouTube.generateVideoDisplay(link);
-			this.message = this.message.replace(/&ab_channel=(.*)(&|)/ig, '').replace(/https:\/\/www\./ig, '');
+		const Spotify = new SpotifyInterface();
+		if (YouTube.linkRegex.test(link) || Spotify.linkRegex.test(link)) {
+			if (YouTube.linkRegex.test(link)) {
+				buf = await YouTube.generateVideoDisplay(link);
+				this.message = this.message.replace(/&ab_channel=(.*)(&|)/ig, '').replace(/https:\/\/www\./ig, '');
+			} else {
+				buf = Spotify.generateSongDisplay(link);
+			}
```
Comment (AnnikaCodes) [inline server/chat-commands/info.ts:2643]:
```suggestion
		if (YouTube.linkRegex.test(link)) {
			buf = await YouTube.generateVideoDisplay(link);
			this.message = this.message.replace(/&ab_channel=(.*)(&|)/ig, '').replace(/https:\/\/www\./ig, '');
		} else if (Spotify.linkRegex.test(link)) {
			buf = Spotify.generateSongDisplay(link);
		}
```

**Review by AnnikaCodes — CHANGES_REQUESTED** [review]

File: server/chat-plugins/the-studio.ts [hunk 3]
```diff
@@ -212,6 +212,29 @@ export class LastFMInterface {
 	}
 }
 
+export class SpotifyInterface {
```
Comment (mia-pi-git) [inline server/chat-plugins/the-studio.ts:215]: Why is this class necessary? Just make it a regex, and use `RegExp.exec` to pull out the song ID.

Comment (mia-pi-git) [reply server/chat-plugins/the-studio.ts:215]: `/show` can handle the song display, and the regex can just be a `const`.

File: server/chat-commands/info.ts (hunk 1 above)
Comment (mia-pi-git) [reply server/chat-commands/info.ts:2584]: No, this doesn't use any sort of fetcher, just a tag.

File: server/chat-plugins/the-studio.ts (hunk 3 above)
Comment (mia-pi-git) [reply server/chat-plugins/the-studio.ts:215]: if you have, say, https://open.spotify.com/track/1SUJo2OxP9QfEegvoXxILL?si=-ybc9PMBTH2u9phSvWekXA.
All you'd need (which would negate the use of this class) would be to just use 
```ts
const spotifyRegex = /open\.spotify\.com\/track\/(.*)/;
const possibleTrack = spotifyRegex.exec(target)?.[0];
if (possibleTrack) {
	buf += `<spotify src="https://open.spotify.com/embed/track/${possibleTrack}"></spotify>`;
} else ....
```

The regex will pull `1SUJo2OxP9QfEegvoXxILL?si=-ybc9PMBTH2u9phSvWekXA`, for you, and as a result can handle both the "is this a spotify url?" and "what's the song id?"

File: server/chat-commands/info.ts (hunk 1 above)
Comment (sirKiraUzumaki) [reply server/chat-commands/info.ts:2584]: Not sure about that.

**Review by mia-pi-git — CHANGES_REQUESTED** [review]
Left the regexes you need to handle this the best. Perhaps make them a top level `const`?

File: server/chat-plugins/the-studio.ts (hunk 3 above)
Comment (sirKiraUzumaki) [reply server/chat-plugins/the-studio.ts:215]: You got a point, lemme just change after checking.

Comment (sirKiraUzumaki) [reply server/chat-plugins/the-studio.ts]: Bte that regex didn't pull out `5h5eqopAnR9BuTdPerTkcM?si=A7KPiDlGQvuJKJMiao0q0g`

Comment (AnnikaCodes) [reply server/chat-plugins/the-studio.ts]: You should do `spotifyRegex.exec(target)?.[1];`, but yes, it works.
image

File: server/chat-commands/info.ts [hunk 4]
```diff
@@ -2622,12 +2630,15 @@ export const commands: ChatCommands = {
...
-
 		let buf;
 		const YouTube = new YoutubeInterface();
+		const Spotify = new SpotifyInterface();
 		if (YouTube.linkRegex.test(link)) {
 			buf = await YouTube.generateVideoDisplay(link);
 			this.message = this.message.replace(/&ab_channel=(.*)(&|)/ig, '').replace(/https:\/\/www\./ig, '');
+		} else if (Spotify.linkRegex.test(link)) {
+			buf = Spotify.generateSongDisplay(link);
+		}
```
Comment (AnnikaCodes) [inline server/chat-commands/info.ts:2641]:
```suggestion
```

File: server/chat-commands/info.ts [hunk 5]
```diff
@@ -2569,11 +2572,16 @@ export const commands: ChatCommands = {
...
+		const Spotify = new SpotifyInterface();
+		const YouTube = new YoutubeInterface();
+		if (Spotify.linkRegex.test(request.link)) {
+			buf = Spotify.generateSongDisplay(request.link);
+		} else if (YouTube.linkRegex.test(request.link)) {
 			buf = await YouTube.generateVideoDisplay(request.link);
 			if (!buf) return this.errorReply('Could not get YouTube video');
 		}
+		}
```
Comment (AnnikaCodes) [inline server/chat-commands/info.ts:2584]:
```suggestion
```

Comment (sirKiraUzumaki) [reply server/chat-commands/info.ts:2584]: No need for suggestions, I am just removing SpotifyInterface class and doing what mia suggested.

File: server/chat-commands/info.ts [hunk 6]
```diff
@@ -2622,12 +2628,17 @@ export const commands: ChatCommands = {
...
-			buf = await YouTube.generateVideoDisplay(link);
-			this.message = this.message.replace(/&ab_channel=(.*)(&|)/ig, '').replace(/https:\/\/www\./ig, '');
+		const SpotifylinkRegex = /open\.spotify\.com\/track\/(.*)/;
+		const SongId = SpotifylinkRegex.exec(link)?.[0];
+		if (YouTube.linkRegex.test(link) || SpotifylinkRegex.test(link)) {
+			if (YouTube.linkRegex.test(link)) {
+				buf = await YouTube.generateVideoDisplay(link);
+				this.message = this.message.replace(/&ab_channel=(.*)(&|)/ig, '').replace(/https:\/\/www\./ig, '');
+			} else {
+				buf = `<spotify src="${SongId}"></spotify>`;
```
Comment (mia-pi-git) [inline server/chat-commands/info.ts:2640]:
```ts
		const spotifyRegex = /open\.spotify\.com\/track\/(.*)/;
```
This should probably be at the top level.
```suggestion
		const songId = spotifyRegex.exec(link)?.[0];
		if (YouTube.linkRegex.test(link)) {
			buf = await YouTube.generateVideoDisplay(link);
			this.message = this.message.replace(/&ab_channel=(.*)(&|)/ig, '').replace(/https:\/\/www\./ig, '');
		} else if (songId) {
				buf = `<spotify src="${songId}"></spotify>`;
```

File: server/chat-commands/info.ts [hunk 7]
```diff
@@ -2521,7 +2521,8 @@ export const commands: ChatCommands = {
 		if (!/^https?:\/\//.test(link)) link = `https://${link}`;
 		link = encodeURI(link);
 		let dimensions;
-		if (!/^(https?:\/\/)?(www\.)?(youtube\.com|youtu\.be)(\/|$)/i.test(link)) {
+		const SpotifyRegex = /open\.spotify\.com\/track\/(.*)/;
```
Comment (mia-pi-git) [inline server/chat-commands/info.ts:2524]: See my below suggestion - just use the top level regex for this.

File: server/chat-commands/info.ts (hunk 6 above)
Comment (mia-pi-git) [reply server/chat-commands/info.ts:2640]:
```ts
		if (YouTube.linkRegex.test(link) || SpotifylinkRegex.test(link)) {
```
(this isn't necessary when you can just use an else-if, as i put in my above suggestion)

File: server/chat-commands/info.ts (hunk 7 above)
Comment (sirKiraUzumaki) [reply server/chat-commands/info.ts:2524]: Where should I define the regex?

Comment (mia-pi-git) [reply server/chat-commands/info.ts:2524]: Put it at the top level of the file.

Comment (mia-pi-git) [reply server/chat-commands/info.ts:2524]: Also reminder to use camelCase, so this should be `spotifyRegex`

File: server/chat-commands/info.ts [hunk 8]
```diff
@@ -2622,12 +2628,15 @@ export const commands: ChatCommands = {
...
-
 		let buf;
 		const YouTube = new YoutubeInterface();
+		const songId = spotifyRegex.exec(link)?.[0];
 		if (YouTube.linkRegex.test(link)) {
 			buf = await YouTube.generateVideoDisplay(link);
 			this.message = this.message.replace(/&ab_channel=(.*)(&|)/ig, '').replace(/https:\/\/www\./ig, '');
+		} else if (songId) {
+				buf = `<spotify src="${songId}"></spotify>`;
+			}
```
Comment (mia-pi-git) [inline server/chat-commands/info.ts:2639]:
```suggestion
		}
```

File: server/chat-commands/info.ts (hunk 7 above)
Comment (AnnikaCodes) [reply server/chat-commands/info.ts:2524]: if it's a top level constant `SPOTIFY_REGEX` would probably be better

**Comment by Zarel** [comment]
The conversation in Dev says that we don't really want Spotify embeds. YouTube also has songs, and unlike Spotify, YouTube's songs are available to everyone, not just Spotify users.

**Comment by sirKiraUzumaki** [comment]
> The conversation in Dev says that we don't really want Spotify embeds. YouTube also has songs, and unlike Spotify, YouTube's songs are available to everyone, not just Spotify users.

Oops I  didn't check this message and keep updating my pr.

**Pull request closed without merging**